core/src/main/java/org/infinispan/persistence/PersistenceUtil.java

```diff
@@ -37,26 +38,30 @@ public static KeyFilter notNull(KeyFilter filter) {
       return filter == null ? KeyFilter.ACCEPT_ALL_FILTER : filter;
    }
 
-   public static <K, V> int count(AdvancedCacheLoader<K, V> acl, KeyFilter<? super K> filter) {
-      final AtomicInteger result = new AtomicInteger(0);
-      acl.process(filter, (marshalledEntry, taskContext) -> result.incrementAndGet(), new WithinThreadExecutor(), false, false);
-      return result.get();
+   public static <K, V> int count(AdvancedCacheLoader<K, V> acl, Predicate<? super K> filter) {
+
+      // This can't be null
+      Long result = Flowable.fromPublisher(acl.publishEntries(filter, false, false)).count().blockingGet();
+      if (result > Integer.MAX_VALUE) {
+         return Integer.MAX_VALUE;
+      }
+      return result.intValue();
    }
 
-   public static <K, V> Set<K> toKeySet(AdvancedCacheLoader<K, V> acl, KeyFilter<? super K> filter) {
+   public static <K, V> Set<K> toKeySet(AdvancedCacheLoader<K, V> acl, Predicate<? super K> filter) {
       if (acl == null)
          return Collections.emptySet();
-      final Set<K> set = new HashSet<K>();
-      acl.process(filter, (marshalledEntry, taskContext) -> set.add(marshalledEntry.getKey()), new WithinThreadExecutor(), false, false);
-      return set;
+      return Flowable.fromPublisher(acl.publishEntries(filter, false, false))
```

Comment:
would a separate publishKeys method make sense?

Reply:
I was actually thinking about that over the weekend. I was debating about having a new method or possibly adding a "builder" API (but this might be overkill).

Something like

```
AdvancedCacheLoader {
   ...
   StorePublisher publish();
   ...
}

interface StorePublisher {
   Publisher<K> keys();

  Publisher<MarshalledEntry<K, V>> entries(boolean fetchValues, boolean fetchKeys);
}
```

What do you think?

Reply:
I wouldn't scope it under such interface, just
```
Publisher<K> publishKeys(Predicate<? super K> filter);
Publisher<MarshalledEntry<K, V>> publishEntries(Predicate<? super K> filter, boolean fetchValue, boolean fetchMetadata);
```
I have considered dropping the `fetchValue` completely but there might be use case for key + metadata (e.g. expiration).

Reply:
Yeah I am fine with the methods as well - and obviously I can make it a nice default method :)

I am thinking that `publishKeys` will have to be documented saying it is up to the cache store implementation whether or not expired keys are returned.

In regards to `fetchValue` I can look over the usages again and see if we can possibly drop it. Although to be honest with you most implementations if you supply either `fetchValue` or `fetchMetadata` it is just as easy to get the other. In that case I wonder if it makes any sense to have these arguments anymore.

Reply:
I guess scattered cache requires just the metadata :) So I guess I will keep both arguments for now.

Reply:
Actually looking at this more `publishKeys` has to guarantee that it returns non expired keys. Or else the usefulness of the method is quite questionable (for example keySet iterator can't use it otherwise). So in that case I will try to limit perf impact as much as possible.